# [概念](https://www.runoob.com/mongodb/mongodb-databases-documents-collections.html)

In [ ]:
SQL术语/概念	MongoDB术语/概念	解释/说明
database	database	数据库
table	collection	数据库表/集合
row	document	数据记录行/文档
column	field	数据字段/域
index	index	索引
table joins	 	表连接,MongoDB不支持
primary key	primary key	主键,MongoDB自动将_id字段设置为主键

下表列出了 RDBMS 与 MongoDB 对应的术语：

RDBMS	MongoDB
数据库	数据库
表格	集合
行	文档
列	字段
表联合	嵌入文档
主键	主键 (MongoDB 提供了 key 为 _id )


数据库服务和客户端
Mysqld/Oracle	mongod
mysql/sqlplus	mongo

## pymongo

查看 mongod 服务是否启动：

In [ ]:
ps aux | grep -v grep | grep mongod
$/usr/local/opt/mongodb/bin/mongod --config /usr/local/etc/mongod.conf

In [ ]:
cd /usr/local/opt/mongodb/bin 
./mongo

数据库也通过名字来标识。数据库名可以是满足以下条件的任意UTF-8字符串。

不能是空字符串（"")。
不得含有' '（空格)、.、$、/、\和\0 (空字符)。
应全部小写。
最多64字节。

### 文档(Document)


文档是一组键值(key-value)对(即 BSON)。

文档中的键/值对是有序的。
文档中的值不仅可以是在双引号里面的字符串，还可以是其他几种数据类型（甚至可以是整个嵌入的文档)。
MongoDB区分类型和大小写。
MongoDB的文档不能有重复的键。
文档的键是字符串。除了少数例外情况，键可以使用任意UTF-8字符。

文档键命名规范：

键不能含有\0 (空字符)。这个字符用来表示键的结尾。
.和$有特别的意义，只有在特定环境下才能使用。
以下划线"_"开头的键是保留的(不是严格要求的)。

### 集合

集合就是 MongoDB 文档组


当第一个文档插入时，集合就会被创建。

合法的集合名
集合名不能是空字符串""。
集合名不能含有\0字符（空字符)，这个字符表示集合名的结尾。
集合名不能以"system."开头，这是为系统集合保留的前缀。
用户创建的集合名字不能含有保留字符。有些驱动程序的确支持在集合名里面包含，这是因为某些系统生成的集合中包含该字符。除非你要访问这种系统创建的集合，否则千万不要在名字里出现$。　

在MongoDB数据库中名字空间 <dbname>.system.* 是包含多种系统信息的特殊集合(Collection)，如下:

集合命名空间	描述
dbname.system.namespaces	列出所有名字空间。
dbname.system.indexes	列出所有索引。
dbname.system.profile	包含数据库概要(profile)信息。
dbname.system.users	列出所有可访问数据库的用户。
dbname.local.sources	包含复制对端（slave）的服务器信息和状态。
对于修改系统集合中的对象有如下限制。

在{{system.indexes}}插入数据，可以创建索引。但除此之外该表信息是不可变的(特殊的drop index命令将自动更新相关信息)。

{{system.users}}是可修改的。 {{system.profile}}是可删除的。

In [ ]:
ObjectId
ObjectId 类似唯一主键，可以很快的去生成和排序，包含 12 bytes，含义是：

前 4 个字节表示创建 unix 时间戳,格林尼治时间 UTC 时间，比北京时间晚了 8 个小时
接下来的 3 个字节是机器标识码
紧接的两个字节由进程 id 组成 PID
最后三个字节是随机数

### init

In [115]:
from bson.objectid import ObjectId
from bson.json_util import loads 
from bson.json_util import dumps 

import pandas as pd
import numpy as np
import pymongo
import json

In [10]:
import pymongo
#mon=pymongo.Connection('localhost',port=27017)
#mongod
import datetime
from pymongo import MongoClient

client = MongoClient('localhost', 27017)

db = client['test-database']
collection = db['test-collection']

In [11]:
post = {"author": "Brettlv",
         "text": "My first blog post!",
         "tags": ["mongodb", "python", "pymongo"],
         "date": datetime.datetime.utcnow()}

posts = db.posts
post_id = posts.insert_one(post).inserted_id

In [51]:
mycollist = db.list_collection_names()

In [52]:
mycollist

['posts', 'test-collection']

In [48]:
mydict = { "name": "RUNOOB", "alexa": "10000", "url": "https://www.runoob.com" }
collection.insert_one(mydict)

In [49]:
for x in collection.find():
    print(x)

{'_id': ObjectId('612cbe6cfabe476a2a90b270'), 'name': 'RUNOOB', 'alexa': '10000', 'url': 'https://www.runoob.com'}


In [50]:
import pprint
pprint.pprint(posts.find_one())

{'_id': ObjectId('59ed769f49963f98f7059270'),
 'author': 'Mike',
 'date': datetime.datetime(2017, 10, 23, 4, 56, 35, 301000),
 'tags': ['mongodb', 'python', 'pymongo'],
 'text': 'My first blog post!'}


In [24]:
dblist = client.list_database_names()
print(dblist)
# dblist = client.database_names() 
if "runoobdb" in dblist:
    print("数据库已存在！")

['CLAGN', 'admin', 'local', 'test-database']


### mydb_CLAGN

In [55]:
mydb_CLAGN = client["CLAGN"]

mycol_CLAGN = mydb_CLAGN["sites"]

collist_CLAGN = mydb_CLAGN.list_collection_names()
# collist = mydb.collection_names()
if "sites" in collist:   # 判断 sites 集合是否存在
    print("集合已存在！")

集合已存在！


In [56]:
mydb_CLAGN.collection_names(include_system_collections=False)

<ipython-input-56-e7d8a59bbb9e>:1: DeprecationWarning: collection_names is deprecated. Use list_collection_names instead.
  mydb_CLAGN.collection_names(include_system_collections=False)


['sites']

### load_sources

In [124]:
CLAGN_statistic_2021_liu=pd.read_excel('/Users/brettlv/blog/IHEP_sync/data/data_clagn/CLAGN_with_RA_Dec_Liuhaov2.xlsx',sheet_name='statistic_paper',header=None)
#CLAGN_list_2021_liu=CLAGN_list_2020[1:]
CLAGN_statistic_2021_liu.columns=['Name','z','type1or2','type','ref1','logM','ref2']
#CLAGN_statistic_2021_liu

In [129]:
CLAGN_statistic_2021_liu_select=CLAGN_statistic_2021_liu[['Name','z','type','logM']]
CLAGN_statistic_2021_liu_select['Note']='Liuetal2021'
zarten_CLAGN_statistic = json.loads(CLAGN_statistic_2021_liu_select.T.to_json()).values()

<ipython-input-129-5470e04a0dfb>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  CLAGN_statistic_2021_liu_select['Note']='Liuetal2021'


zarten_CLAGN_statistic

In [131]:
mycol_CLAGN.insert_many(zarten_CLAGN_statistic)

### insert data

In [57]:
mydict_1

{'name': 'Mrk1018',
 'title': 'Long-term and multiwavelength evolution of a changing-look AGN Mrk 1018',
 'ref': '2021MNRAS.506.4188L',
 '_id': ObjectId('612cbe25fabe476a2a90b26f')}

mydict_1 = { "name": "Mrk1018", "title": "Long-term and multiwavelength evolution of a changing-look AGN Mrk 1018", "ref": "2021MNRAS.506.4188L" }
x = mycol_CLAGN.insert_one(mydict_1) 
print(x)
x = mycol_CLAGN.find_one()
print(x)

mydict_2={"_id": 0,'url':'https://ui.adsabs.harvard.edu/abs/2021MNRAS.506.4188L/abstract'}
x = mycol_CLAGN.insert_one(mydict_2) 
print(x)

insert_many()来执行批量插入操作

new_posts = [{"_id": 1000,
               "author": "Curry",
               "text": "Another post!",
               "tags": ["bulk", "insert"],
               "date": datetime.datetime(2017, 11, 12, 11, 14)},
              {"_id": 1001,"author": "Maxsu",
               "title": "MongoDB is fun",
               "text": "and pretty easy too!",
               "date": datetime.datetime(2019, 11, 10, 10, 45)}]

posts = db.posts
result = posts.insert_many(new_posts)
print("Bulk Inserts Result is :", result.inserted_ids)
#print (post)

### find

In [ ]:
from bson.objectid import ObjectId

# The web framework gets post_id from the URL and passes it as a string
def get(post_id):
    # Convert from string to ObjectId:
    document = client.db.collection.find_one({'_id': ObjectId(post_id)})
    
#//原文出自【易百教程】，商业转载请联系作者获得授权，非商业请保留原文链接：https://www.yiibai.com/mongodb/mongodb_python.html

#### 正则

In [82]:
myquery = {"name": {"$regex": "^R"}} #以R开头

for x in mycol_CLAGN.find({},{ "_id": 0, "name": 1, "url": 1 }):
    print(x)

{'name': 'Mrk1018', 'url': 'https://ui.adsabs.harvard.edu/abs/2021MNRAS.506.4188L'}
{'url': 'https://ui.adsabs.harvard.edu/abs/2021MNRAS.506.4188L/abstract'}
{'url': 'https://ui.adsabs.harvard.edu/abs/2021MNRAS.506.4188L/abstract'}


除了 _id 你不能在一个对象中同时指定 0 和 1，如果你设置了一个字段为 0，则其他都为 1，反之亦然。

In [85]:
for x in mycol_CLAGN.find({},{ "url": 0 }):
    print(x)

{'_id': ObjectId('612cbe25fabe476a2a90b26f'), 'name': 'Mrk1018', 'title': 'Long-term and multiwavelength evolution of a changing-look AGN Mrk 1018', 'ref': '2021MNRAS.506.4188L'}
{'_id': '0'}
{'_id': 0}


读取 name 字段中第一个字母 ASCII 值大于 "H" 的数据，大于的修饰符条件为 {"$gt": "H"} :

In [86]:
myquery_2 = { "name": { "$gt": "H" } }
mydoc = mycol_CLAGN.find(myquery_2)
 
for x in mydoc:
    print(x)

{'_id': ObjectId('612cbe25fabe476a2a90b26f'), 'name': 'Mrk1018', 'title': 'Long-term and multiwavelength evolution of a changing-look AGN Mrk 1018', 'ref': '2021MNRAS.506.4188L', 'url': 'https://ui.adsabs.harvard.edu/abs/2021MNRAS.506.4188L'}


d = datetime.datetime(2019, 11, 12, 12)
for post in posts.find({"date": {"$lt": d}}).sort("author"):
    pprint.pprint(post)

### print

In [134]:
for x in mycol_CLAGN.find({'Note':'Liuetal2021'}):
    print(x)

{'_id': ObjectId('612d0315fabe476a2a90b274'), 'Name': '1ES 1927+654 ', 'z': 0.017, 'type': ' A ', 'logM': 7.3, 'Note': 'Liuetal2021'}
{'_id': ObjectId('612d0315fabe476a2a90b275'), 'Name': '1H 0419-577 ', 'z': 0.104, 'type': ' B ', 'logM': 8.58, 'Note': 'Liuetal2021'}
{'_id': ObjectId('612d0315fabe476a2a90b276'), 'Name': '3C 390.3 ', 'z': 0.056, 'type': ' A ', 'logM': ' - ', 'Note': 'Liuetal2021'}
{'_id': ObjectId('612d0315fabe476a2a90b277'), 'Name': 'ESO 362-G18 ', 'z': 0.012, 'type': ' A ', 'logM': 7.65, 'Note': 'Liuetal2021'}
{'_id': ObjectId('612d0315fabe476a2a90b278'), 'Name': 'Fairall 9 ', 'z': 0.046, 'type': ' A ', 'logM': 8.41, 'Note': 'Liuetal2021'}
{'_id': ObjectId('612d0315fabe476a2a90b279'), 'Name': 'HE 1136-2304 ', 'z': 0.027, 'type': ' A ', 'logM': 7.58, 'Note': 'Liuetal2021'}
{'_id': ObjectId('612d0315fabe476a2a90b27a'), 'Name': 'IC 751 ', 'z': 0.032, 'type': ' B ', 'logM': 8.5, 'Note': 'Liuetal2021'}
{'_id': ObjectId('612d0315fabe476a2a90b27b'), 'Name': 'IRAS 23226-3843 

In [76]:
myresult_limit3 = mycol_CLAGN.find().limit(1)
# 输出结果
for x in myresult_limit3:
    print(x)

{'_id': ObjectId('612cbe25fabe476a2a90b26f'), 'name': 'Mrk1018', 'title': 'Long-term and multiwavelength evolution of a changing-look AGN Mrk 1018', 'ref': '2021MNRAS.506.4188L', 'url': 'https://ui.adsabs.harvard.edu/abs/2021MNRAS.506.4188L'}


In [80]:
myquery_1 = { "name":  "Mrk1018" }
mydoc = mycol_CLAGN.find(myquery_1)
for x in mydoc:
    print(x)

{'_id': ObjectId('612cbe25fabe476a2a90b26f'), 'name': 'Mrk1018', 'title': 'Long-term and multiwavelength evolution of a changing-look AGN Mrk 1018', 'ref': '2021MNRAS.506.4188L', 'url': 'https://ui.adsabs.harvard.edu/abs/2021MNRAS.506.4188L'}


### delete

myquery_1 = { "name": {"$regex": "^R"}}#删除所有 name 字段中以 F 开头的文档:
x = mycol_CLAGN.delete_many(myquery_1)
print(x.deleted_count, "个文档已删除")

In [67]:
myquery_2 = { "_id": '0'}#删除所有 name 字段中以 F 开头的文档:
x = mycol_CLAGN.delete_many(myquery_2)
print(x.deleted_count, "个文档已删除")

1 个文档已删除


myquery = { "name": "Taobao" }
mycol.delete_one(myquery)

myquery = { "name": {"$regex": "^F"} }
x = mycol.delete_many(myquery)
print(x.deleted_count, "个文档已删除")

mycol.drop()# drop() 方法来删除一个集合

### edit data

In [65]:
myquery = { "ref": "2021MNRAS.506.4188L" }
newvalues = { "$set": { "url": "https://ui.adsabs.harvard.edu/abs/2021MNRAS.506.4188L" } }
mycol_CLAGN.update_one(myquery, newvalues)

In [101]:
mydoc = mycol_CLAGN.find().sort("name",-1)#第二个字段指定排序规则，1 为升序，-1 为降序，默认为升序
for x in mydoc:
    print(x)

{'_id': 0, 'url': 'https://ui.adsabs.harvard.edu/abs/2020ApJ...892...93C', 'name': 'NGC4395', 'title': 'Variability and the Size-Luminosity Relation of the Intermediate-mass AGN in NGC 4395'}
{'_id': '0', 'url': 'https://ui.adsabs.harvard.edu/abs/2021MNRAS.507..687J', 'name': 'NGC1566', 'title': 'Broad-band X-ray observations of the 2018 outburst of the changing-look active galactic nucleus NGC 1566'}
{'_id': ObjectId('612cbe25fabe476a2a90b26f'), 'name': 'Mrk1018', 'title': 'Long-term and multiwavelength evolution of a changing-look AGN Mrk 1018', 'ref': '2021MNRAS.506.4188L', 'url': 'https://ui.adsabs.harvard.edu/abs/2021MNRAS.506.4188L'}


In [98]:
myquery = { "_id": "0" }
newvalues = { "$set": { "name": "NGC1566" } }
mycol_CLAGN.update_one(myquery, newvalues)

myquery = { "_id": "0" }
newvalues = { "$set": { "title": "Broad-band X-ray observations of the 2018 outburst of the changing-look active galactic nucleus NGC 1566" } }
mycol_CLAGN.update_one(myquery, newvalues)

myquery = { "_id": "0" }
newvalues = { "$set": { "url": "https://ui.adsabs.harvard.edu/abs/2021MNRAS.507..687J" } }
mycol_CLAGN.update_one(myquery, newvalues)





In [99]:
myquery = { "_id": 0 }
newvalues = { "$set": { "name": "NGC4395" } }
mycol_CLAGN.update_one(myquery, newvalues)


myquery = { "_id": 0 }
newvalues = { "$set": { "title": "Variability and the Size-Luminosity Relation of the Intermediate-mass AGN in NGC 4395" } }
mycol_CLAGN.update_one(myquery, newvalues)

myquery = { "_id": 0 }
newvalues = { "$set": { "url": "https://ui.adsabs.harvard.edu/abs/2020ApJ...892...93C" } }
mycol_CLAGN.update_one(myquery, newvalues)


In [102]:
for x in mycol_CLAGN.find():
    print(x)

{'_id': ObjectId('612cbe25fabe476a2a90b26f'), 'name': 'Mrk1018', 'title': 'Long-term and multiwavelength evolution of a changing-look AGN Mrk 1018', 'ref': '2021MNRAS.506.4188L', 'url': 'https://ui.adsabs.harvard.edu/abs/2021MNRAS.506.4188L'}
{'_id': '0', 'url': 'https://ui.adsabs.harvard.edu/abs/2021MNRAS.507..687J', 'name': 'NGC1566', 'title': 'Broad-band X-ray observations of the 2018 outburst of the changing-look active galactic nucleus NGC 1566'}
{'_id': 0, 'url': 'https://ui.adsabs.harvard.edu/abs/2020ApJ...892...93C', 'name': 'NGC4395', 'title': 'Variability and the Size-Luminosity Relation of the Intermediate-mass AGN in NGC 4395'}


myquery = { "address": { "$regex": "^S" } }
newvalues = { "$set": { "name": "Minnie" } }
x = mycol.update_many(myquery, newvalues)


### count

计数统计如果只想知道有多少文档匹配查询，可以执行count()方法操作

In [ ]:
print("posts count is = ", posts.count())
print("posts's author is Maxsu count is =", posts.find({"author": "Maxsu"}).count())

## pandas

### 读取mongdb数据

In [106]:
import pandas as pd 
# 设置MongoDB连接信息

data = pd.DataFrame(list(mycol_CLAGN.find()))
# 删除mongodb中的_id字段del data['_id']#
选择需要显示的字段data = data[['name','title','url']]
#print(data)

data

In [111]:
import pymongo
import pandas as pd
import csv 
df_data = mycol_CLAGN.find({ 'name':{'$regex':'^N'}}) 

In [113]:
for i in df_data:
    print(i)

{'_id': '0', 'url': 'https://ui.adsabs.harvard.edu/abs/2021MNRAS.507..687J', 'name': 'NGC1566', 'title': 'Broad-band X-ray observations of the 2018 outburst of the changing-look active galactic nucleus NGC 1566'}
{'_id': 0, 'url': 'https://ui.adsabs.harvard.edu/abs/2020ApJ...892...93C', 'name': 'NGC4395', 'title': 'Variability and the Size-Luminosity Relation of the Intermediate-mass AGN in NGC 4395'}


### 写入mongo

In [116]:
zarten_df = pd.DataFrame(np.arange(12).reshape((3,4)))
zarten_data = json.loads(zarten_df.T.to_json()).values()
#collection.insert_many(zarten_data)

In [117]:
zarten_df 

0  1   2   3
0  0  1   2   3
1  4  5   6   7
2  8  9  10  11

In [119]:
zarten_df.T

0  1   2
0  0  4   8
1  1  5   9
2  2  6  10
3  3  7  11

In [118]:
zarten_data

dict_values([{'0': 0, '1': 1, '2': 2, '3': 3}, {'0': 4, '1': 5, '2': 6, '3': 7}, {'0': 8, '1': 9, '2': 10, '3': 11}])

In [120]:
mycol.insert_many(zarten_data)

In [123]:
for i in mycol.find():
    print(i)

{'_id': ObjectId('612cbe25fabe476a2a90b26f'), 'name': 'Mrk1018', 'title': 'Long-term and multiwavelength evolution of a changing-look AGN Mrk 1018', 'ref': '2021MNRAS.506.4188L', 'url': 'https://ui.adsabs.harvard.edu/abs/2021MNRAS.506.4188L'}
{'_id': '0', 'url': 'https://ui.adsabs.harvard.edu/abs/2021MNRAS.507..687J', 'name': 'NGC1566', 'title': 'Broad-band X-ray observations of the 2018 outburst of the changing-look active galactic nucleus NGC 1566'}
{'_id': 0, 'url': 'https://ui.adsabs.harvard.edu/abs/2020ApJ...892...93C', 'name': 'NGC4395', 'title': 'Variability and the Size-Luminosity Relation of the Intermediate-mass AGN in NGC 4395'}
{'_id': ObjectId('612d0151fabe476a2a90b271'), '0': 0, '1': 1, '2': 2, '3': 3}
{'_id': ObjectId('612d0151fabe476a2a90b272'), '0': 4, '1': 5, '2': 6, '3': 7}
{'_id': ObjectId('612d0151fabe476a2a90b273'), '0': 8, '1': 9, '2': 10, '3': 11}


## sqlite3

In [ ]:
sqlite3是内置模块，所以不需要安装的，直接import导入即可：

In [ ]:
import sqlite3
query="""
CREATE TABLE test
(a VARCHAR(20),b VARCHAR(20),c REAL, d INTEGER);
"""
con=sqlite3.connect(':memory:')
con.execute(query)
con.commit()


data={('altanta','georgia',1.25,6),('takkahassee','florida',2.6,3),('sacramento','california',1.7,5)}
stmt="INSERT INTO test VALUES(?,?,?,?)"
con.executemany(stmt,data)
con.commit()

cursor=con.execute('select * from test')
rows =cursor.fetchall()

rows

cursor.description



## pandas.io.sql

In [ ]:
import pandas
from pandas import DataFrame
#DataFrame(rows,columns=zip(*cursor.description)[0])

import pandas.io.sql as sql
sql.read_sql('select * from test',con)
sql.read_sql_query('select * from test',con)

## pymysql

In [ ]:
PyMySQL是一个python与mysql的沟通管道，需要你在本地安装配置好mysql才能使用

In [ ]:
import pymysql.cursors

conn = pymysql.connect(host='localhost', port=3306, user='root', passwd='', db='mysql')

cur = conn.cursor()

cur.execute("SELECT Host,User FROM user")

print(cur.description)

print()

for row in cur:
    print(row)

cur.close()
conn.close()

In [ ]:
import pymysql.cursors



# Connect to the database
connection = pymysql.connect(host='localhost',
                             user='user',
                             password='passwd',
                             db='db',
                             charset='utf8mb4',
                             cursorclass=pymysql.cursors.DictCursor)

try:
    with connection.cursor() as cursor:
        # Create a new record
        sql = "INSERT INTO `users` (`email`, `password`) VALUES (%s, %s)"
        cursor.execute(sql, ('webmaster@python.org', 'very-secret'))

    # connection is not autocommit by default. So you must commit to save
    # your changes.
    connection.commit()

    with connection.cursor() as cursor:
        # Read a single record
        sql = "SELECT `id`, `password` FROM `users` WHERE `email`=%s"
        cursor.execute(sql, ('webmaster@python.org',))
        result = cursor.fetchone()
        print(result)
finally:
    connection.close()

## MySQLdb

In [ ]:
为了用DB-API编写MySQL脚本，必须确保已经安装了MySQL

In [8]:
import MySQLdb

ModuleNotFoundError: No module named 'MySQLdb'